src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Protocol/Completion/RazorVSInternalCompletionParams.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Text.Json.Serialization;
+
+namespace Roslyn.LanguageServer.Protocol;
+
+/// <summary>
+/// A replacement of the LSP protocol <see cref="CompletionParams"/> that ensures correct serialization between LSP servers.
+/// </summary>
+/// <remarks>
+/// This is the same as the LSP protocol <see cref="CompletionParams"/> except that it strongly types the <see cref="Context"/> property as VSInternalCompletionContext,
+/// because our custom message target gets handled by a JsonRpc connection set up by the editor, that has no Roslyn converters.
+/// Without using this class, we lose VSInternalCompletionContext.InvokeKind property when calling Roslyn or HTML servers from Razor
+/// which results in default value of "Invoked" to be used and can cause overly aggressive completion list.
+/// </remarks>
+internal sealed class RazorVSInternalCompletionParams : TextDocumentPositionParams, IPartialResultParams<SumType<CompletionItem[], CompletionList>?>, IWorkDoneProgressOptions
```

Comment:
🤔 I've been thinking for a bit about whether this should be called `RazorCompletionParams` since there's no such type as `VSInternalCompletionParams`, but I think I'm OK with it. Using this name as an indicator that it provides a `VSInternalCompletionContext` seems a strong enough reason.

Reply:
That's pretty much exactly the logic I used trying to decide between those exact two names :)